HMR / dev-server broken since first iteration of of assets-webpack-plugin

Hi everybody,

I have troubles using the dev-server since https://github.com/symfony/webpack-encore/commit/2f106e88f122
The following problem occurs with and without HMR enabled (`encore dev-server` / `encore dev-server --hot` makes no difference):

The initial compilation is running fine. Any subsequent change of a file crashes the dev server. The stack trace points to https://github.com/symfony/webpack-encore/blob/c2f386f0a64d374730d680fb149f0234dc700d7a/lib/plugins/entry-files-manifest.js#L30, which seems to be the culprit here. Putting a `console.log(assets[asset][fileType]);` right before that line shows that there's a structural difference in the payload that's received by this plugin between the initial compilation and subsequent, partial compilations.

Current versions and logs can be found below.
Let me know if you need some more information about the things I do in my snippets here, about my development env or something else :)

Best,
Alex

---

OS: macOS 10.13.6
Node Version: 8.11.3
NPM Version: 6.4.1
webpack-encore version: master @ c2f386f0a64d3
webpack-dev-server 3.1.10
webpack 4.25.0

### Output

```
npm run dev-server

> @myproject@ dev-server /.../myproject
> encore dev-server --hot

Running webpack-dev-server ...

 DONE  Compiled successfully in 7750ms                                                                                                                                    13:58:47

 WAIT  Compiling...                                                                                                                                                       13:59:10

/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32
            assets[asset][fileType] = assets[asset][fileType].map(buildPath => buildPath.replace(/^\//g, ''));
                                                                                         ^

TypeError: buildPath.replace is not a function
    at assets.(anonymous function).(anonymous function).assets.(anonymous function).(anonymous function).map.buildPath (/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32:90)
    at Array.map (<anonymous>)
    at Object.processOutput (/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32:63)
    at /.../myproject/node_modules/assets-webpack-plugin/dist/lib/output/createOutputWriter.js:67:30
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:511:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @myproject@ dev-server: `encore dev-server --hot`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @myproject@ dev-server script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

### Output with log:
```
Running webpack-dev-server ...

[ 'http://localhost:8080/build/runtime.js',
  'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/app~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/app~js/pages/deal/deal-edit.js',
  'http://localhost:8080/build/vendors~js/app.js',
  'http://localhost:8080/build/css/app~js/app.js',
  'http://localhost:8080/build/js/app.js' ]
[ 'http://localhost:8080/build/vendors~js/app.css',
  'http://localhost:8080/build/css/app~js/app.css',
  'http://localhost:8080/build/js/app.css' ]
[ 'http://localhost:8080/build/runtime.js',
  'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/app~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~chartModuleDeps~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/placement~js/review.js',
  'http://localhost:8080/build/js/placement~js/review.js',
  'http://localhost:8080/build/js/review.js' ]
[ 'http://localhost:8080/build/js/placement~js/review.css',
  'http://localhost:8080/build/js/review.css' ]
[ 'http://localhost:8080/build/runtime.js',
  'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/app~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~chartModuleDeps~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/placement~js/review.js',
  'http://localhost:8080/build/js/placement~js/review.js',
  'http://localhost:8080/build/js/placement.js' ]
[ 'http://localhost:8080/build/js/placement~js/review.css',
  'http://localhost:8080/build/js/placement.css' ]
[ 'http://localhost:8080/build/runtime.js',
  'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
  'http://localhost:8080/build/vendors~js/app~js/pages/deal/deal-edit.js',
  'http://localhost:8080/build/js/pages/deal/deal-edit.js' ]
[ 'http://localhost:8080/build/runtime.js',
  'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
  'http://localhost:8080/build/css/app~js/app.js' ]
[ 'http://localhost:8080/build/css/app~js/app.css' ]
 DONE  Compiled successfully in 7695ms                                                                                                                                    14:01:56

 WAIT  Compiling...                                                                                                                                                       14:15:42

[ { js: 
     [ 'http://localhost:8080/build/runtime.js',
       'http://localhost:8080/build/vendors~css/app~js/app~js/pages/deal/deal-edit~js/placement~js/review.js',
       'http://localhost:8080/build/vendors~js/app~js/placement~js/review.js',
       'http://localhost:8080/build/vendors~js/app~js/pages/deal/deal-edit.js',
       'http://localhost:8080/build/vendors~js/app.js',
       'http://localhost:8080/build/css/app~js/app.js',
       'http://localhost:8080/build/js/app.js' ],
    css: 
     [ 'http://localhost:8080/build/vendors~js/app.css',
       'http://localhost:8080/build/css/app~js/app.css',
       'http://localhost:8080/build/js/app.css' ] } ]
/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32
            assets[asset][fileType] = assets[asset][fileType].map(buildPath => buildPath.replace(/^\//g, ''));
                                                                                         ^

TypeError: buildPath.replace is not a function
    at assets.(anonymous function).(anonymous function).assets.(anonymous function).(anonymous function).map.buildPath (/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32:90)
    at Array.map (<anonymous>)
    at Object.processOutput (/.../myproject/node_modules/@symfony/webpack-encore/lib/plugins/entry-files-manifest.js:32:63)
    at /.../myproject/node_modules/assets-webpack-plugin/dist/lib/output/createOutputWriter.js:67:30
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:511:3)
```

### webpack.config.js

```
const Encore = require('@symfony/webpack-encore');

Encore
// the project directory where compiled assets will be stored
    .setOutputPath('public/build/')

    // the public path used by the web server to access the previous directory
    .setPublicPath('/build')

    .cleanupOutputBeforeBuild()

    .enableSourceMaps(!Encore.isProduction())

    // create hashed filenames (e.g. app.abc123.css)
    .enableVersioning(Encore.isProduction())

    // js entrypoints / chunks
    .addEntry('js/app', './assets/js/app.js')
    .addEntry('js/review', './assets/js/review.js')
    .addEntry('js/placement', './assets/js/placement.js')
    .addEntry('js/pages/deal/deal-edit', './assets/js/pages/deal/deal-edit.js')

    // scss entrypoints / chunks
    .addStyleEntry('css/app', './assets/css/app.scss')

    // use Sass/SCSS files
    .enableSassLoader()

    // allow .vue files
    .enableVueLoader()

    // autoprovision for legacy applications that require $/jQuery as a global variable
    .autoProvidejQuery()

    // empty the outputPath dir before each build
    .cleanupOutputBeforeBuild()

    // show OS notifications when builds finish/fail
    .enableBuildNotifications()

    .splitEntryChunks()

    .enableSingleRuntimeChunk();

// export the final configuration
module.exports = Encore.getWebpackConfig();

--- a/lib/plugins/entry-files-manifest.js
+++ b/lib/plugins/entry-files-manifest.js
@@ -19,6 +19,13 @@ function processOutput(assets) {
         delete assets[entry];
     }
 
+    // with --watch or dev-server, subsequent calls will include
+    // the original assets (so, assets.entrypoints) + the new
+    // assets (which will have their original structure). We
+    // delete the entrypoints key, and then process the new assets
+    // like normal below
+    delete assets.entrypoints;
+
     // This will iterate over all the entry points and remove the / from the start of the paths. It also converts the
     // one file entries into an array of one entry since that was how the entry point file was before this change.
     for (const asset in assets) {

